lib/default.nix

```
@@ -375,8 +351,149 @@ let
           default = { };
           description = "A non-block device";
         };
+        _meta = lib.mkOption {
+          internal = true;
+          description = ''
+            meta informationen generated by disko
+            currently used for building a dependency list so we know in which order to create the devices
+          '';
+          default = diskoLib.deepMergeMap (dev: dev._meta) (flatten (map attrValues (attrValues devices)));
+        };
+        _packages = lib.mkOption {
+          internal = true;
+          description = ''
+            packages required by the disko configuration
+          '';
+          default = pkgs: unique (flatten (map (dev: dev._pkgs pkgs) (flatten (map attrValues (attrValues devices)))));
+        };
+        _scripts = lib.mkOption {
+          internal = true;
+          description = ''
+            The scripts generated by disko
+          '';
+          default = { pkgs, checked ? false }: {
+            umountScript = (diskoLib.writeCheckedBash { inherit pkgs checked; }) "disko-umount" ''
+              export PATH=${lib.makeBinPath (with pkgs; [
+                util-linux
+                e2fsprogs
+                mdadm
+                zfs
+                lvm2
+              ])}:$PATH
+              ${cfg.config._umount}
+            '';
+
+            formatScript = (diskoLib.writeCheckedBash { inherit pkgs checked; }) "disko-create" ''
+              export PATH=${lib.makeBinPath (cfg.config._packages pkgs)}:$PATH
+              ${cfg.config._create}
+            '';
+
+            mountScript = (diskoLib.writeCheckedBash { inherit pkgs checked; }) "disko-mount" ''
+              export PATH=${lib.makeBinPath (cfg.config._packages pkgs)}:$PATH
+              ${cfg.config._mount}
+            '';
+
+            diskoScript = (diskoLib.writeCheckedBash { inherit pkgs checked; }) "disko" ''
+              export PATH=${lib.makeBinPath ((cfg.config._packages pkgs) ++ [ pkgs.bash ])}:$PATH
+              ${cfg.config._disko}
+            '';
+
+            # These are useful to skip copying executables uploading a script to an in-memory installer
+            umountScriptNoDeps = (diskoLib.writeCheckedBash { inherit pkgs checked; noDeps = true; }) "disko-umount" ''
+              ${cfg.config._umount}
+            '';
+
+            formatScriptNoDeps = (diskoLib.writeCheckedBash { inherit pkgs checked; noDeps = true; }) "disko-create" ''
+              ${cfg.config._create}
+            '';
+
+            system.build.mountScriptNoDeps = (diskoLib.writeCheckedBash { inherit pkgs checked; noDeps = true; }) "disko-mount" ''
+              ${cfg.config._mount}
+            '';
+
+            diskoScriptNoDeps = (diskoLib.writeCheckedBash { inherit pkgs checked; noDeps = true; }) "disko" ''
+              ${cfg.config._disko}
+            '';
+          };
+        };
+        _umount = lib.mkOption {
```

Would it not make sense to make this more scarier?

```suggestion
        _umountAndDestroy = lib.mkOption {
```

Otherwise I would actually call it _destroy instead of _umount.

---

ah, I wanted to change it to just umount, the destroying is basically an undesired side effect :) but for now it would make sense yes